File: api-kt/src/main/kotlin/site/daydream/mote/service/DriveUploadService.kt
package site.daydream.mote.service

import org.springframework.dao.support.DataAccessUtils
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.support.TransactionTemplate
import site.daydream.mote.config.UploadConfig
import site.daydream.mote.exception.BadRequestException
import site.daydream.mote.exception.ConflictException
import site.daydream.mote.exception.NotFoundException
import site.daydream.mote.model.DriveNode
import site.daydream.mote.model.DriveUpload
import site.daydream.mote.model.DriveUploadInitRequest
import java.io.File
import java.io.InputStream
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.security.MessageDigest

@Service
class DriveUploadService(
    private val jdbc: NamedParameterJdbcTemplate,
    private val driveService: DriveService,
    private val uploadConfig: UploadConfig,
    txManager: PlatformTransactionManager,
) {
    private val tx = TransactionTemplate(txManager)

    private fun chunksDir(uploadId: String): File =
        Paths.get(uploadConfig.uploadDir, "drive", "_chunks", uploadId).toFile()

    fun init(req: DriveUploadInitRequest): DriveUpload {
        DriveService.validateName(req.name)
        if (req.size <= 0 || req.size > MAX_FILE_SIZE) throw BadRequestException("file too large")
        var chunk = if (req.chunkSize <= 0) DEFAULT_CHUNK else req.chunkSize
        if (chunk < (1L shl 20) || chunk > (64L shl 20)) throw BadRequestException("invalid chunk size")
        if (req.parentId != null) driveService.requireActiveFolder(req.parentId)

        val total = ((req.size + chunk - 1) / chunk).toInt()
        val mask = ByteArray((total + 7) / 8)
        val id = DriveService.randomHex(16)
        val now = System.currentTimeMillis()
        val expiresAt = now + UPLOAD_TTL_MS

        Files.createDirectories(chunksDir(id).toPath())
        try {
            jdbc.update(
                """
                INSERT INTO drive_uploads (id, parent_id, name, size, chunk_size, total_chunks, received_mask, status, expires_at, created_at, updated_at)
                VALUES (:id, :pid, :name, :size, :ch, :tot, :mask, 'uploading', :exp, :now, :now)
                """.trimIndent(),
                mapOf(
                    "id" to id, "pid" to req.parentId, "name" to req.name,
                    "size" to req.size, "ch" to chunk, "tot" to total, "mask" to mask,
                    "exp" to expiresAt, "now" to now,
                ),
            )
        } catch (e: Exception) {
            chunksDir(id).deleteRecursively()
            throw e
        }
        return findOrThrow(id)
    }

    fun get(id: String): Pair<DriveUpload, List<Int>> {
        val u = findOrThrow(id)
        return u to decodeMask(u.receivedMask, u.totalChunks)
    }

    fun putChunk(id: String, idx: Int, body: InputStream) {
        val u = findOrThrow(id)
        if (u.status != "uploading") throw BadRequestException("invalid upload state")
        if (idx < 0 || idx >= u.totalChunks) throw BadRequestException("chunk index out of range")

        val expected = if (idx == u.totalChunks - 1) u.size - idx.toLong() * u.chunkSize else u.chunkSize
        val tmp = File(chunksDir(id), "$idx.part")
        val final = File(chunksDir(id), "$idx.bin")

        // Copy up to expected+1 bytes to detect oversize.
        var written = 0L
        tmp.outputStream().use { out ->
            val buf = ByteArray(64 * 1024)
            var remaining = expected + 1
            while (remaining > 0) {
                val toRead = minOf(buf.size.toLong(), remaining).toInt()
                val n = body.read(buf, 0, toRead)
                if (n < 0) break
                out.write(buf, 0, n)
                written += n
                remaining -= n
            }
        }
        if (written != expected) {
            tmp.delete()
            throw BadRequestException("chunk size mismatch")
        }
        Files.move(tmp.toPath(), final.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)

        // Update bitmap atomically inside a serial transaction.
        tx.execute {
            val cur = jdbc.queryForObject(
                "SELECT received_mask FROM drive_uploads WHERE id = :id",
                mapOf("id" to id), ByteArray::class.java,
            ) ?: throw NotFoundException("upload not found")
            if (idx / 8 >= cur.size) throw BadRequestException("invalid chunk index")
            cur[idx / 8] = (cur[idx / 8].toInt() or (1 shl (idx % 8))).toByte()
            jdbc.update(
                "UPDATE drive_uploads SET received_mask = :m, updated_at = :now WHERE id = :id",
                mapOf("m" to cur, "now" to System.currentTimeMillis(), "id" to id),
            )
        }
    }

    fun complete(id: String, onCollision: String): DriveNode {
        val u = findOrThrow(id)
        val updated = jdbc.update(
            """
            UPDATE drive_uploads SET status = 'assembling', updated_at = :now
            WHERE id = :id AND status = 'uploading'
            """.trimIndent(),
            mapOf("now" to System.currentTimeMillis(), "id" to id),
        )
        if (updated == 0) throw BadRequestException("invalid upload state")

        try {
            val mask = jdbc.queryForObject(
                "SELECT received_mask FROM drive_uploads WHERE id = :id",
                mapOf("id" to id), ByteArray::class.java,
            ) ?: throw NotFoundException("upload not found")
            if (mask.size * 8 < u.totalChunks) throw BadRequestException("invalid upload state")
            for (i in 0 until u.totalChunks) {
                if (mask[i / 8].toInt() and (1 shl (i % 8)) == 0) {
                    throw BadRequestException("upload incomplete")
                }
            }

            if (u.parentId != null) driveService.requireActiveFolder(u.parentId)

            var finalName = u.name
            val existing = driveService.findActiveSibling(u.parentId, finalName)
            if (existing != null) {
                when (onCollision) {
                    "skip" -> {
                        deleteSession(id)
                        return existing
                    }
                    "rename" -> finalName = driveService.autoRename(u.parentId, finalName)
                    "overwrite" -> {
                        // handled below by replaceFileNode
                    }
                    else -> throw ConflictException("name already exists in this folder")
                }
            }

            val blobName = DriveService.newBlobName(finalName)
            val relPath = "drive/$blobName"
            val absPath = Paths.get(uploadConfig.uploadDir, relPath).toFile()
            val tmpAbs = File(absPath.absolutePath + ".part")

            val (hash, written) = try {
                assemble(u, tmpAbs)
            } catch (e: Exception) {
                tmpAbs.delete()
                throw e
            }
            if (written != u.size) {
                tmpAbs.delete()
                throw BadRequestException("final file size mismatch")
            }
            try {
                Files.move(tmpAbs.toPath(), absPath.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
            } catch (e: Exception) {
                tmpAbs.delete()
                throw e
            }

            val node = try {
                if (onCollision == "overwrite") {
                    driveService.replaceFileNode(u.parentId, finalName, relPath, hash, u.size)
                } else {
                    driveService.createFileNode(u.parentId, finalName, relPath, hash, u.size)
                }
            } catch (e: Exception) {
                absPath.delete()
                throw e
            }

            deleteSession(id)
            return node
        } catch (e: Exception) {
            // Allow client to retry.
            jdbc.update(
                "UPDATE drive_uploads SET status = 'uploading' WHERE id = :id",
                mapOf("id" to id),
            )
            throw e
        }
    }

    fun cancel(id: String) {
        val n = jdbc.update(
            "DELETE FROM drive_uploads WHERE id = :id AND status = 'uploading'",
            mapOf("id" to id),
        )
        if (n > 0) chunksDir(id).deleteRecursively()
    }

    fun purgeExpired(): Int {
        val ids = jdbc.queryForList(
            "SELECT id FROM drive_uploads WHERE expires_at < :now",
            mapOf("now" to System.currentTimeMillis()),
            String::class.java,
        )
        ids.forEach { deleteSession(it) }
        return ids.size
    }

    private fun deleteSession(id: String) {
        jdbc.update("DELETE FROM drive_uploads WHERE id = :id", mapOf("id" to id))
        chunksDir(id).deleteRecursively()
    }

    private fun assemble(u: DriveUpload, out: File): Pair<String, Long> {
        val md = MessageDigest.getInstance("SHA-256")
        var total = 0L
        out.outputStream().use { os ->
            val buf = ByteArray(64 * 1024)
            for (i in 0 until u.totalChunks) {
                val cf = File(chunksDir(u.id), "$i.bin")
                cf.inputStream().use { ins ->
                    while (true) {
                        val r = ins.read(buf)
                        if (r < 0) break
                        os.write(buf, 0, r)
                        md.update(buf, 0, r)
                        total += r
                    }
                }
            }
            os.flush()
        }
        val sb = StringBuilder()
        for (b in md.digest()) sb.append("%02x".format(b.toInt() and 0xff))
        return sb.toString() to total
    }

    private fun findOrThrow(id: String): DriveUpload {
        val u = DataAccessUtils.singleResult(jdbc.query(
            "SELECT * FROM drive_uploads WHERE id = :id",
            mapOf("id" to id),
        ) { rs, _ ->
            DriveUpload(
                id = rs.getString("id"),
                parentId = rs.getObject("parent_id")?.let { (it as Number).toLong() },
                name = rs.getString("name"),
                size = rs.getLong("size"),
                chunkSize = rs.getLong("chunk_size"),
                totalChunks = rs.getInt("total_chunks"),
                receivedMask = rs.getBytes("received_mask") ?: ByteArray(0),
                status = rs.getString("status"),
                expiresAt = rs.getLong("expires_at"),
                createdAt = rs.getLong("created_at"),
                updatedAt = rs.getLong("updated_at"),
            )
        }) ?: throw NotFoundException("upload not found")
        if (u.expiresAt < System.currentTimeMillis()) throw NotFoundException("upload expired")
        return u
    }

    companion object {
        private const val UPLOAD_TTL_MS = 24L * 60 * 60 * 1000
        private const val MAX_FILE_SIZE = 4L shl 30
        private const val DEFAULT_CHUNK = 8L shl 20

        fun decodeMask(mask: ByteArray, total: Int): List<Int> {
            val out = ArrayList<Int>()
            for (i in 0 until total) {
                if (i / 8 >= mask.size) break
                if ((mask[i / 8].toInt() and (1 shl (i % 8))) != 0) out.add(i)
            }
            return out
        }
    }
}

File: api-rs/src/service/drive_upload_service.rs
use crate::config::UploadConfig;
use crate::model::drive::{
    DriveNode, DriveUpload, DriveUploadInitRequest,
};
use crate::service::drive_service::{
    new_blob_name, new_token, valid_name, DriveError, DriveResult, DriveService,
};
use chrono::Utc;
use sha2::Digest;
use sqlx::SqlitePool;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use tokio::io::{AsyncReadExt, AsyncWriteExt};

const UPLOAD_TTL_MS: i64 = 24 * 60 * 60 * 1000;
const MAX_FILE_SIZE: i64 = 4 << 30;
const DEFAULT_CHUNK_SIZE: i64 = 8 << 20;

pub struct DriveUploadService {
    pub pool: SqlitePool,
    pub drive: Arc<DriveService>,
    pub config: UploadConfig,
}

impl DriveUploadService {
    pub fn new(pool: SqlitePool, drive: Arc<DriveService>, config: UploadConfig) -> Arc<Self> {
        Arc::new(Self { pool, drive, config })
    }

    fn chunks_dir(&self, upload_id: &str) -> PathBuf {
        Path::new(&self.config.base_path)
            .join("drive")
            .join("_chunks")
            .join(upload_id)
    }

    pub async fn init(&self, req: &DriveUploadInitRequest) -> DriveResult<DriveUpload> {
        valid_name(&req.name)?;
        if req.size <= 0 || req.size > MAX_FILE_SIZE {
            return Err(DriveError::UploadTooLarge);
        }
        let chunk = if req.chunk_size <= 0 {
            DEFAULT_CHUNK_SIZE
        } else {
            req.chunk_size
        };
        if !(1 << 20..=64 << 20).contains(&chunk) {
            return Err(DriveError::UploadInvalidRequest);
        }
        if let Some(pid) = req.parent_id {
            self.drive.require_active_folder(pid).await?;
        }
        let total = ((req.size + chunk - 1) / chunk) as i64;
        let mask = vec![0u8; ((total + 7) / 8) as usize];
        let id = new_token(16);
        let now = Utc::now().timestamp_millis();
        let expires = now + UPLOAD_TTL_MS;

        tokio::fs::create_dir_all(self.chunks_dir(&id)).await?;

        let res = sqlx::query(
            "INSERT INTO drive_uploads (id, parent_id, name, size, chunk_size, total_chunks, received_mask, status, expires_at, created_at, updated_at)
             VALUES (?, ?, ?, ?, ?, ?, ?, 'uploading', ?, ?, ?)",
        )
        .bind(&id)
        .bind(req.parent_id)
        .bind(&req.name)
        .bind(req.size)
        .bind(chunk)
        .bind(total)
        .bind(&mask)
        .bind(expires)
        .bind(now)
        .bind(now)
        .execute(&self.pool)
        .await;
        if let Err(e) = res {
            let _ = tokio::fs::remove_dir_all(self.chunks_dir(&id)).await;
            return Err(DriveError::Sqlx(e));
        }
        self.find(&id).await
    }

    pub async fn find(&self, id: &str) -> DriveResult<DriveUpload> {
        let u: Option<DriveUpload> =
            sqlx::query_as("SELECT * FROM drive_uploads WHERE id = ?")
                .bind(id)
                .fetch_optional(&self.pool)
                .await?;
        let u = u.ok_or(DriveError::UploadNotFound)?;
        if u.expires_at < Utc::now().timestamp_millis() {
            return Err(DriveError::UploadNotFound);
        }
        Ok(u)
    }

    pub async fn get_status(&self, id: &str) -> DriveResult<(DriveUpload, Vec<i64>)> {
        let u = self.find(id).await?;
        let received = decode_mask(&u.received_mask, u.total_chunks);
        Ok((u, received))
    }

    pub async fn put_chunk(
        &self,
        id: &str,
        idx: i64,
        mut body: impl tokio::io::AsyncRead + Unpin,
    ) -> DriveResult<()> {
        let u = self.find(id).await?;
        if u.status != "uploading" {
            return Err(DriveError::UploadInvalidRequest);
        }
        if idx < 0 || idx >= u.total_chunks {
            return Err(DriveError::UploadChunkOOR);
        }
        let expected = if idx == u.total_chunks - 1 {
            u.size - idx * u.chunk_size
        } else {
            u.chunk_size
        };

        let dir = self.chunks_dir(id);
        let tmp = dir.join(format!("{}.part", idx));
        let final_p = dir.join(format!("{}.bin", idx));

        let mut f = tokio::fs::File::create(&tmp).await?;
        let mut buf = vec![0u8; 64 * 1024];
        let mut written: i64 = 0;
        let limit = expected + 1;
        loop {
            let to_read = std::cmp::min(buf.len() as i64, limit - written) as usize;
            if to_read == 0 {
                break;
            }
            let n = body.read(&mut buf[..to_read]).await?;
            if n == 0 {
                break;
            }
            f.write_all(&buf[..n]).await?;
            written += n as i64;
        }
        f.flush().await?;
        drop(f);

        if written != expected {
            let _ = tokio::fs::remove_file(&tmp).await;
            return Err(DriveError::UploadChunkSize);
        }
        tokio::fs::rename(&tmp, &final_p).await?;

        // Atomic mask OR via SQL.
        let byte_idx = (idx / 8) as i64;
        let bit_mask = 1i64 << (idx % 8);
        let now = Utc::now().timestamp_millis();
        // SQLite has no bitwise blob operators, but we can extract the byte and update.
        // Simplest correct approach: read+modify+write under BEGIN IMMEDIATE.
        let mut conn = self.pool.acquire().await?;
        sqlx::query("BEGIN IMMEDIATE").execute(&mut *conn).await?;
        let mask: Vec<u8> = match sqlx::query_scalar::<_, Vec<u8>>(
            "SELECT received_mask FROM drive_uploads WHERE id = ?",
        )
        .bind(id)
        .fetch_one(&mut *conn)
        .await
        {
            Ok(m) => m,
            Err(e) => {
                let _ = sqlx::query("ROLLBACK").execute(&mut *conn).await;
                return Err(DriveError::Sqlx(e));
            }
        };
        let mut mask = mask;
        if (byte_idx as usize) >= mask.len() {
            let _ = sqlx::query("ROLLBACK").execute(&mut *conn).await;
            return Err(DriveError::UploadInvalidRequest);
        }
        mask[byte_idx as usize] |= bit_mask as u8;
        sqlx::query("UPDATE drive_uploads SET received_mask = ?, updated_at = ? WHERE id = ?")
            .bind(&mask)
            .bind(now)
            .bind(id)
            .execute(&mut *conn)
            .await?;
        sqlx::query("COMMIT").execute(&mut *conn).await?;
        Ok(())
    }

    pub async fn complete(
        &self,
        id: &str,
        on_collision: &str,
    ) -> DriveResult<DriveNode> {
        let u = self.find(id).await?;
        let now = Utc::now().timestamp_millis();
        let res = sqlx::query(
            "UPDATE drive_uploads SET status = 'assembling', updated_at = ?
             WHERE id = ? AND status = 'uploading'",
        )
        .bind(now)
        .bind(id)
        .execute(&self.pool)
        .await?;
        if res.rows_affected() == 0 {
            return Err(DriveError::UploadInvalidRequest);
        }

        // Verify mask
        let mask: Vec<u8> =
            sqlx::query_scalar("SELECT received_mask FROM drive_uploads WHERE id = ?")
                .bind(id)
                .fetch_one(&self.pool)
                .await?;
        if (mask.len() as i64) * 8 < u.total_chunks {
            self.mark_uploading(id).await;
            return Err(DriveError::UploadInvalidRequest);
        }
        for i in 0..u.total_chunks {
            if mask[(i / 8) as usize] & (1 << (i % 8)) == 0 {
                self.mark_uploading(id).await;
                return Err(DriveError::UploadIncomplete);
            }
        }

        let parent_id = u.parent_id;
        if let Some(pid) = parent_id {
            if let Err(e) = self.drive.require_active_folder(pid).await {
                self.mark_uploading(id).await;
                return Err(e);
            }
        }
        let mut final_name = u.name.clone();
        let existing = self
            .drive
            .find_active_sibling(parent_id, &final_name)
            .await?;
        if existing.is_some() {
            match on_collision {
                "skip" => {
                    self.delete_session(id).await;
                    return Ok(existing.unwrap());
                }
                "rename" => {
                    final_name = self.drive.auto_rename(parent_id, &final_name).await?;
                }
                "overwrite" => {}
                _ => {
                    self.mark_uploading(id).await;
                    return Err(DriveError::NameConflict);
                }
            }
        }

        let blob_name = new_blob_name(&final_name);
        let rel_path = format!("drive/{}", blob_name);
        let abs_path = self.drive.blob_abs_path(&rel_path);
        let tmp_abs = abs_path.with_extension(format!(
            "{}.part",
            abs_path
                .extension()
                .map(|s| s.to_string_lossy().into_owned())
                .unwrap_or_default()
        ));

        let assembled = self.assemble(&u, &tmp_abs).await;
        let (hash, written) = match assembled {
            Ok(v) => v,
            Err(e) => {
                let _ = tokio::fs::remove_file(&tmp_abs).await;
                self.mark_uploading(id).await;
                return Err(e);
            }
        };
        if written != u.size {
            let _ = tokio::fs::remove_file(&tmp_abs).await;
            self.mark_uploading(id).await;
            return Err(DriveError::UploadFinalSize);
        }
        if let Err(e) = tokio::fs::rename(&tmp_abs, &abs_path).await {
            let _ = tokio::fs::remove_file(&tmp_abs).await;
            self.mark_uploading(id).await;
            return Err(DriveError::Io(e));
        }

        let node_res = if on_collision == "overwrite" {
            self.drive
                .replace_file_node(parent_id, &final_name, &rel_path, &hash, u.size)
                .await
        } else {
            self.drive
                .create_file_node(parent_id, &final_name, &rel_path, &hash, u.size)
                .await
        };
        match node_res {
            Ok(node) => {
                self.delete_session(id).await;
                Ok(node)
            }
            Err(e) => {
                let _ = tokio::fs::remove_file(&abs_path).await;
                self.mark_uploading(id).await;
                Err(e)
            }
        }
    }

    pub async fn cancel(&self, id: &str) -> DriveResult<()> {
        let res = sqlx::query("DELETE FROM drive_uploads WHERE id = ? AND status = 'uploading'")
            .bind(id)
            .execute(&self.pool)
            .await?;
        if res.rows_affected() == 0 {
            return Ok(());
        }
        let _ = tokio::fs::remove_dir_all(self.chunks_dir(id)).await;
        Ok(())
    }

    pub async fn purge_expired(&self) -> DriveResult<usize> {
        let now = Utc::now().timestamp_millis();
        let ids: Vec<(String,)> =
            sqlx::query_as("SELECT id FROM drive_uploads WHERE expires_at < ?")
                .bind(now)
                .fetch_all(&self.pool)
                .await?;
        let count = ids.len();
        for (id,) in ids {
            self.delete_session(&id).await;
        }
        Ok(count)
    }

    async fn delete_session(&self, id: &str) {
        let _ = sqlx::query("DELETE FROM drive_uploads WHERE id = ?")
            .bind(id)
            .execute(&self.pool)
            .await;
        let _ = tokio::fs::remove_dir_all(self.chunks_dir(id)).await;
    }

    async fn mark_uploading(&self, id: &str) {
        let _ = sqlx::query("UPDATE drive_uploads SET status = 'uploading' WHERE id = ?")
            .bind(id)
            .execute(&self.pool)
            .await;
    }

    async fn assemble(&self, u: &DriveUpload, out_path: &Path) -> DriveResult<(String, i64)> {
        let mut out = tokio::fs::File::create(out_path).await?;
        let mut hasher = sha2::Sha256::new();
        let mut total: i64 = 0;
        let dir = self.chunks_dir(&u.id);
        let mut buf = vec![0u8; 64 * 1024];
        for i in 0..u.total_chunks {
            let p = dir.join(format!("{}.bin", i));
            let mut f = tokio::fs::File::open(&p).await?;
            loop {
                let n = f.read(&mut buf).await?;
                if n == 0 {
                    break;
                }
                hasher.update(&buf[..n]);
                out.write_all(&buf[..n]).await?;
                total += n as i64;
            }
        }
        out.flush().await?;
        let hash = hex::encode(hasher.finalize());
        Ok((hash, total))
    }
}

fn decode_mask(mask: &[u8], total: i64) -> Vec<i64> {
    let mut out = Vec::with_capacity(total as usize);
    for i in 0..total {
        let byte = (i / 8) as usize;
        if byte >= mask.len() {
            break;
        }
        if mask[byte] & (1 << (i % 8)) != 0 {
            out.push(i);
        }
    }
    out
}
